core/src/test/java/kafka/test/annotation/ClusterTest.java

```diff
@@ -34,6 +35,7 @@
 @Target({METHOD})
 @Retention(RUNTIME)
 @TestTemplate
+@Timeout(30)
```

Comment:
In some case, the timeout is 120 and the other is 30.
Should we set it to 120 to avoid to make test flaky?

Reply:
Hello @TaiJuWu,
In this PR, we're setting a 30-second timeout limit for all new test infrastructure. We’ve already identified and fixed test cases exceeding this limit in a [previous PR](https://github.com/apache/kafka/pull/16796) and addressed these issues in a follow-up [PR](https://github.com/apache/kafka/pull/16810). Moving forward, developers should ensure their tests complete within this 30-second window to avoid prolonged execution times.
WDYT?

Reply:
Does this work? I know we have tried setting global timeouts in the past with no success.

Also, 30 seconds might be too short for `ClusterTest` tests since these are integration tests. I think 120 is more reasonable.

Can you add a new test that just sleeps for 5 minutes to verify the behavior?

Reply:
> Does this work? I know we have tried setting global timeouts in the past with no success.

yes, it works. see the QA (https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-16957/3/tests) and following error:
```
Error
java.util.concurrent.TimeoutException: testClientInstanceId(kafka.test.ClusterInstance) timed out after 30 seconds
Stacktrace
java.util.concurrent.TimeoutException: testClientInstanceId(kafka.test.ClusterInstance) timed out after 30 seconds
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:647)
	at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:272)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1384)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	at java.util.ArrayList.forEach(ArrayList.java:1259)
	Suppressed: org.apache.kafka.common.errors.InterruptException: java.lang.InterruptedException
		at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.maybeThrowInterruptException(ConsumerNetworkClient.java:537)
		at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:298)
		at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:253)
		at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.poll(ConsumerNetworkClient.java:244)
		at org.apache.kafka.clients.consumer.internals.ConsumerNetworkClient.awaitMetadataUpdate(ConsumerNetworkClient.java:166)
		at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureCoordinatorReady(AbstractCoordinator.java:303)
		at org.apache.kafka.clients.consumer.internals.AbstractCoordinator.ensureCoordinatorReady(AbstractCoordinator.java:264)
		at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.coordinatorUnknownAndUnreadySync(ConsumerCoordinator.java:451)
		at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.commitOffsetsSync(ConsumerCoordinator.java:1112)
		at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.maybeAutoCommitOffsetsSync(ConsumerCoordinator.java:1144)
		at org.apache.kafka.clients.consumer.internals.ConsumerCoordinator.close(ConsumerCoordinator.java:980)
		at org.apache.kafka.clients.consumer.internals.ClassicKafkaConsumer.lambda$close$3(ClassicKafkaConsumer.java:1140)
		at org.apache.kafka.common.utils.Utils.swallow(Utils.java:1050)
		at org.apache.kafka.clients.consumer.internals.ClassicKafkaConsumer.close(ClassicKafkaConsumer.java:1140)
		at org.apache.kafka.clients.consumer.internals.ClassicKafkaConsumer.close(ClassicKafkaConsumer.java:1109)
		at org.apache.kafka.clients.consumer.internals.ClassicKafkaConsumer.close(ClassicKafkaConsumer.java:1097)
		at org.apache.kafka.clients.consumer.KafkaConsumer.close(KafkaConsumer.java:1757)
		at kafka.admin.ClientTelemetryTest.testClientInstanceId(ClientTelemetryTest.java:116)
		at java.lang.reflect.Method.invoke(Method.java:498)
		... 16 more
	Caused by: java.lang.InterruptedException
		... 35 more
```

BTW, it is not a "true" global timeout since it is applied to the tests using `ClusterTest`. That means the tests extending `KafkaServerTestHarness` are not limited.

Reply:
Very nice! 

Is it possible for `@ClusterTest`s to override this annotation for longer timeout?

Reply:
> Hello @TaiJuWu, In this PR, we're setting a 30-second timeout limit for all new test infrastructure. We’ve already identified and fixed test cases exceeding this limit in a [previous PR](https://github.com/apache/kafka/pull/16796) and addressed these issues in a follow-up [PR](https://github.com/apache/kafka/pull/16810). Moving forward, developers should ensure their tests complete within this 30-second window to avoid prolonged execution times. WDYT?

Got it. Sound great. Thanks for your explain.

Reply:
> Is it possible for `@ClusterTests` to override this annotation for longer timeout?

I don't consider I can achieve this using the `@Timeout` annotation alone. Developers should avoid scenarios where they need to modify this setting. If a test genuinely requires more than 30 seconds, the `@Timeout`annotation can be applied at the method level, it will override this annotation.
```java
    @Timeout(1)
    @ClusterTests({
            @ClusterTest(types = {Type.ZK}),
            @ClusterTest(types = {Type.KRAFT}, serverProperties = {
                    @ClusterConfigProperty(key = AUTHORIZER_CLASS_NAME_CONFIG, value = STANDARD_AUTHORIZER)
            })
    })
    public void testAclCliWithAdminAPI(ClusterInstance cluster) throws InterruptedException {
        Thread.sleep(2000);
        testAclCli(cluster, adminArgs(cluster.bootstrapServers(), Optional.empty()));
    }
```